File: services/internal-treasury/src/test/java/br/com/deltaglobalbank/domain/listPix/ListMakePixUseCaseTest.java
package br.com.deltaglobalbank.domain.listPix;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

import br.com.deltaglobalbank.internal_treasury.domain.makePix.MakePix;
import br.com.deltaglobalbank.internal_treasury.domain.makePix.MakePixRepository;
import br.com.deltaglobalbank.internal_treasury.domain.shared.PaymentsStatus;
import br.com.deltaglobalbank.internal_treasury.features.listPix.ListPixResponse;
import br.com.deltaglobalbank.internal_treasury.features.listPix.ListPixUseCase;
import org.junit.jupiter.api.Test;
import org.springframework.data.domain.PageImpl;

class ListMakePixUseCaseTest {

    private final MakePixRepository repository = mock(MakePixRepository.class);
    private final ListPixUseCase useCase = new ListPixUseCase(repository);

    @Test
    void deveRetornarApenasWaitingQuandoFiltroNaoInformado() {
        MakePix transference = new MakePix(
            UUID.randomUUID(), 123456L, "12", "1234", "214421",
            PaymentsStatus.WAITING, "214214", "Teste", 10050L, Instant.now()
        );

        PageImpl<MakePix> page = new PageImpl<>(List.of(transference));
        when(repository.findByStatus(eq(PaymentsStatus.WAITING), any())).thenReturn(page);

        ListPixResponse response = useCase.execute(0, 10, "waiting");

        assertEquals(1, response.content().size());
        assertEquals("WAITING", response.content().get(0).status());
    }

    @Test
    void deveRetornarApprovedQuandoFiltroForApproved() {
        MakePix transference = new MakePix(
            UUID.randomUUID(), 123456L, "12", "1234", "214421",
            PaymentsStatus.APPROVED, "214214", "Teste", 10050L, Instant.now()
        );

        PageImpl<MakePix> page = new PageImpl<>(List.of(transference));
        when(repository.findByStatus(eq(PaymentsStatus.APPROVED), any())).thenReturn(page);

        ListPixResponse response = useCase.execute(0, 10, "approved");

        assertEquals(1, response.content().size());
        assertEquals("APPROVED", response.content().get(0).status());
    }

    @Test
    void deveRetornarDeniedQuandoFiltroForDenied() {
        MakePix transference = new MakePix(
            UUID.randomUUID(), 123456L, "12", "1234", "214421",
            PaymentsStatus.DENIED, "214214", "Teste", 10050L, Instant.now()
        );

        PageImpl<MakePix> page = new PageImpl<>(List.of(transference));
        when(repository.findByStatus(eq(PaymentsStatus.DENIED), any())).thenReturn(page);

        ListPixResponse response = useCase.execute(0, 10, "denied");

        assertEquals(1, response.content().size());
        assertEquals("DENIED", response.content().get(0).status());
    }

    @Test
    void deveRetornarTodosQuandoFiltroForAll() {
        MakePix transference = new MakePix(
            UUID.randomUUID(), 123456L, "12", "1234", "214421",
            PaymentsStatus.WAITING, "214214", "Teste", 10050L, Instant.now()
        );

        PageImpl<MakePix> page = new PageImpl<>(List.of(transference));
        when(repository.findAll(any())).thenReturn(page);

        ListPixResponse response = useCase.execute(0, 10, "all");

        assertEquals(1, response.content().size());
        assertEquals("WAITING", response.content().get(0).status());
    }
}

File: services/internal-treasury/src/test/kotlin/br/com/deltaglobalbank/domain/listPix/ListMakePixUseCaseTest.kt
package br.com.deltaglobalbank.domain.listPix

import br.com.deltaglobalbank.internal_treasury.domain.makePix.MakePix
import br.com.deltaglobalbank.internal_treasury.domain.makePix.MakePixRepository
import br.com.deltaglobalbank.internal_treasury.domain.shared.PaymentsStatus
import br.com.deltaglobalbank.internal_treasury.features.listPix.ListPixUseCase
import org.junit.jupiter.api.Assertions.assertEquals
import org.mockito.Mockito.mock
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.mockito.kotlin.whenever
import org.springframework.data.domain.PageImpl
import java.time.Instant
import java.util.UUID
import kotlin.test.Test

class ListMakePixUseCaseTest {
    private val repository = mock<MakePixRepository>()
    private val useCase = ListPixUseCase(repository)

    @Test
    fun `deve retornar apenas waiting quando filtro nao informado`() {
        val transference = MakePix(
            id = UUID.randomUUID(),
            accountId = 123456L,
            recipientInstitutionCode = "12",
            operationAmount = 10050,
            status = PaymentsStatus.WAITING,
            recipientName = "Teste",
            createdAt = Instant.now(),
            recipientBranchCode = "1234",
            recipientAccountNumber = "214421",
            recipientAccountType = "214214"
        )

        val page = PageImpl(listOf(transference))
        whenever(repository.findByStatus(eq(PaymentsStatus.WAITING), any())).thenReturn(page)

        val response = useCase.execute(0, 10, "waiting")

        assertEquals(1, response.content.size)
        assertEquals("WAITING", response.content[0].status)
    }

    @Test
    fun `deve retornar approved quando filtro for approved`() {
        val transference = MakePix(
            id = UUID.randomUUID(),
            accountId = 123456L,
            recipientInstitutionCode = "12",
            operationAmount = 10050,
            status = PaymentsStatus.APPROVED,
            recipientName = "Teste",
            createdAt = Instant.now(),
            recipientBranchCode = "1234",
            recipientAccountNumber = "214421",
            recipientAccountType = "214214"
        )

        val page = PageImpl(listOf(transference))
        whenever(repository.findByStatus(eq(PaymentsStatus.APPROVED), any())).thenReturn(page)

        val response = useCase.execute(0, 10, "approved")

        assertEquals(1, response.content.size)
        assertEquals("APPROVED", response.content[0].status)
    }

    @Test
    fun `deve retornar denied quando filtro for denied`() {
        val transference = MakePix(
            id = UUID.randomUUID(),
            accountId = 123456L,
            recipientInstitutionCode = "12",
            operationAmount = 10050,
            status = PaymentsStatus.DENIED,
            recipientName = "Teste",
            createdAt = Instant.now(),
            recipientBranchCode = "1234",
            recipientAccountNumber = "214421",
            recipientAccountType = "214214"
        )

        val page = PageImpl(listOf(transference))
        whenever(repository.findByStatus(eq(PaymentsStatus.DENIED), any())).thenReturn(page)

        val response = useCase.execute(0, 10, "denied")

        assertEquals(1, response.content.size)
        assertEquals("DENIED", response.content[0].status)
    }

    @Test
    fun `deve retornar todos quando filtro for all`() {
        val transference = MakePix(
            id = UUID.randomUUID(),
            accountId = 123456L,
            recipientInstitutionCode = "12",
            operationAmount = 10050,
            status = PaymentsStatus.WAITING,
            recipientName = "Teste",
            createdAt = Instant.now(),
            recipientBranchCode = "1234",
            recipientAccountNumber = "214421",
            recipientAccountType = "214214"
        )

        val page = PageImpl(listOf(transference))
        whenever(repository.findAll(any())).thenReturn(page)

        val response = useCase.execute(0, 10, "all")

        assertEquals(1, response.content.size)
        assertEquals("WAITING", response.content[0].status)
    }
}